lib/dry/initializer/builders/attribute.rb

```diff
@@ -72,9 +72,21 @@ def coercion_line
       arity = @type.is_a?(Proc) ? @type.arity : @type.method(:call).arity
 
       if arity.equal?(1) || arity.negative?
-        "#{@val} = #{@item}.type.call(#{@val}) unless #{@null} == #{@val}"
+        <<-METH
+          begin
+            #{@val} = #{@item}.type.call(#{@val}) unless #{@null} == #{@val}
+          rescue Dry::Types::ConstraintError => e
+            raise Dry::Initializer::CoercionError.new(e, '#{@source}')
+          end
+        METH
       else
-        "#{@val} = #{@item}.type.call(#{@val}, self) unless #{@null} == #{@val}"
+        <<-METH
+          begin
+            #{@val} = #{@item}.type.call(#{@val}, self) unless #{@null} == #{@val}
+          rescue Dry::Types::ConstraintError => e
+            raise Dry::Initializer::CoercionError.new(e, '#{@source}')
+          end
+        METH
```

Comment:
how about

```ruby
type_call_params = (arity.equal?(1) || arity.negative?) ? @val : "#{@val}, self"

 <<-METH
  begin
    #{@val} = #{@item}.type.call(#{type_call_params}) unless #{@null} == #{@val}
  rescue Dry::Types::ConstraintError => e
    raise Dry::Initializer::CoercionError.new(e, '#{@source}')
  end
METH
```

Reply:
I took the liberty of implementing this suggestion to clean up the code. Small change to the suggestion to keep the line length below 80. This PR was already approved but not yet merged, hope this is ok.